<div align="center">
    <h1 align="center">Tugas Besar IF2220 Probabilitas dan Statistika 2023/2024</h1>
    <h3>Topik : Health</h3>
    <strong>
        <h4>Dibuat oleh: </h4>
        <h4>K01 - Kelompok 24</h4>
            <table>
              <tr>
                <td>NIM</td>
                <td>Nama</td>
              </tr>
              <tr>
                <td>13522071</td>
                <td>Bagas Sambega Rosyada</td>
              </tr>
              <tr>
                <td>13522091</td>
                <td>Raden Francisco Trianto Bratadiningrat</td>
              </tr>
            </table>
      </strong>
</div>


## Inisialisasi Dependencies Python dan Pembacaan Data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats

# Read data from CSV
data = pd.read_csv("health.csv", index_col=0)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      2000 non-null   int64  
 1   Income                   2000 non-null   float64
 2   Gender                   2000 non-null   object 
 3   Education                2000 non-null   object 
 4   Stress_Level             2000 non-null   float64
 5   Exercise_Hours_Per_Week  2000 non-null   float64
 6   Cholesterol_Level        2000 non-null   float64
 7   Weight                   2000 non-null   float64
 8   Height                   2000 non-null   float64
 9   Blood_Pressure           2000 non-null   float64
 10  Health_Status            2000 non-null   object 
dtypes: float64(7), int64(1), object(3)
memory usage: 187.5+ KB


# Soal General

## Nomor 1

Menulis deskripsi statistika (Descriptive Statistics) dari semua kolom pada data. Data
yang bersifat numerik dapat diberikan nilai mean, median, modus, standar deviasi,
variansi, range, nilai minimum, maksimum, kuartil, IQR, skewness dan kurtosis. Data
dalam bentuk string dapat dicari unique values, dan proporsi nya.

### Jawab: Menulis Deskripsi Statistika dari semua kolom pada data

Kolom-Kolom dibagi menjadi bentuk data numerik dan string:
- String: Gender, Education, Health_Status
- Numerik: Age, Income, Stress_level, Exercise_Hours_Per_Week, Cholesterol_Level, Weight, Height, Blood_Pressure

#### Data String

Untuk data string akan dicari:
- Unique values
- Proporsi

In [2]:
# Get columns with string type
string_columns = data.select_dtypes(include=['object']).columns

# For all colum with type string
for column in string_columns:
    stat = pd.DataFrame()
    
    # Get unique values and their counts
    unique_values = data[column].value_counts()
    
    # Calculate proportions
    proportions = unique_values*100 / len(data)

    # Construct frame
    stat["Unique Value"] = unique_values.index
    stat["Frekuensi"] = unique_values.values
    stat["Proporsi (%)"] = proportions.values
    
    # Print result
    print("Data for " + column)
    print(stat)
    print("\n")

Data for Gender
  Unique Value  Frekuensi  Proporsi (%)
0       Female       1011         50.55
1         Male        989         49.45


Data for Education
        Unique Value  Frekuensi  Proporsi (%)
0  Bachelor's Degree        940         47.00
1    Master's Degree        482         24.10
2        High School        271         13.55
3                PhD        247         12.35
4              Other         57          2.85
5          undefined          3          0.15


Data for Health_Status
  Unique Value  Frekuensi  Proporsi (%)
0         Fair        519         25.95
1         Poor        504         25.20
2    Excellent        503         25.15
3         Good        474         23.70




#### Data Numerik

Untuk data numerik akan dicari:
- mean 
- median
- modus
- standar deviasi
- variansi
- range
- minimum
- maksimum
- kuartil
- IQR
- skewness
- kurtosis

In [3]:
# Set pandas to display without scientific notation
pd.set_option('display.float_format', lambda x: '%.0f' % x)

# Get numeric data
numerik_data = data.select_dtypes(include=['number'])

# Calculate the mode for each column
def get_mode(x: pd.DataFrame) -> pd.Series | None:
    try:
        _ = x.mode()[1]
        mode = 'multivalue'
    except KeyError:
        mode = x.mode()[0]
    return mode

# Construct dataframe
numerik_stat = pd.DataFrame()

numerik_stat["Mean"] = numerik_data.mean().round(3)
numerik_stat["Median"] = numerik_data.median().round(3)
numerik_stat["Modus"] = numerik_data.apply(get_mode).round(3)
numerik_stat["Deviasi"] = numerik_data.std().round(3)
numerik_stat["Variansi"] = numerik_data.var().round(3)
numerik_stat["Min"] = numerik_data.min().round(3)
numerik_stat["Max"] = numerik_data.max().round(3)
numerik_stat["Q1"] = numerik_data.quantile(0.25).round(3)
numerik_stat["Q2"] = numerik_data.quantile(0.5).round(3)
numerik_stat["Q3"] = numerik_data.quantile(0.75).round(3)
numerik_stat["IQR"] = (numerik_stat["Q3"] - numerik_stat["Q1"]).round(3)
numerik_stat["Skewness"] = numerik_data.skew().round(3)
numerik_stat["Kurtosis"] = numerik_data.kurtosis().round(3)

# Print result
numerik_stat

Mean  Median       Modus  Deviasi      Variansi  \
Age                          39      39          53       12           134   
Income                  4889928 4898901  multivalue  2010795 4043296630639   
Stress_Level                  5       5  multivalue        1             1   
Exercise_Hours_Per_Week      10      10  multivalue        5            24   
Cholesterol_Level           200     200  multivalue        1             1   
Weight                       70      70  multivalue       10           104   
Height                      151     151  multivalue        1             0   
Blood_Pressure              120     120           0       11           125   

                             Min      Max      Q1      Q2      Q3     IQR  \
Age                           20       59      30      39      49      19   
Income                  -1088987 12227492 3580714 4898901 6271595 2690882   
Stress_Level                   3        9       5       5       6       1   
Exercise_Hours_Per_Week       -4       27       7      10      13       7   
Cholesterol_Level            197      203     200     200     201       1   
Weight                        33      110      64      70      77      13   
Height                       149      154     150     151     151       1   
Blood_Pressure                 0      201     113     120     127      14   

                         Skewness  Kurtosis  
Age                            -0        -1  
Income                         -0        -0  
Stress_Level                   -0        -0  
Exercise_Hours_Per_Week         0         0  
Cholesterol_Level              -0         0  
Weight                         -0         0  
Height                          1         1  
Blood_Pressure                 -2        19

## Nomor 2

Apakah pada data tersebut terdapat outlier? Jika ya, dapatkah anda menanganinya?
Jelaskan apa yang umumnya dilakukan untuk menangani outlier.

### Jawab

Outlier adalah data yang memiliki nilai jauh dari kumpulan data-data lainnya yang terjadi misalkan karena kegagalan pengukuran, salah masukan, dll.

Berdasarkan proses untuk mengatasi dan menghapus data outlier, kami menemukan data memiliki outlier. Hal ini dapat terlihat karena jumlah data pada hasil proses penghapusan outlier, jumlah data mengurang. Namun hal tersebut juga menunjukkan bahwa kami dapat menangani data-data outlier dengan melakukan penghapusa data outlier sebelum digunakan untuk analisis yang lebih lanjut.

#### Langkah Mengatasi Outlier
#### 1. Menghilangkan data yang mengandung nilai kosong atau null karena dapat merusak analisis

Data yang kosong atau null, akan menggangu serta merusak analis. data dengan nilai kosong akan dihapus.

#### 2. Menghilangkan data yang tidak masuk akal

Batas-batas yang digunakan:
 
- Age: Non-Negative
- Stress_Level: Non-Negative
- Exercise_Hours_Per_Week: Non-Negative
- Cholesterol_Level: Non-Negative
- Weight: Non-Negative
- Height: Non-Negative
- Blood_Pressure: Non-Negative

#### 3. Menggunakan Nilai Z pada Standar Deviasi

Nilai z dari suatu data akan menunjukan seberapa jauh suatu data dengan data lainnya. Secara umum digunakan batas nilai z = 3 sebagai batas apakah suatu data terlalu jauh dari kumpulan data. Dengan menghitung nilai z dari semua kolom yang bertipe numerik, serta menghilangkan data yang memiliki salah satu kolomnya dengan nilai z mutlak lebih besar dari 3, maka kami berhasil mengatasi data-data outlier.



In [4]:
# 1. Remove Null Values 
no_null_data = data.dropna()

# 2. Remove non logical data
result_data = no_null_data.copy()
result_data = result_data[result_data['Age'] >= 0]
result_data = result_data[result_data['Stress_Level'] >= 0]
result_data = result_data[result_data['Exercise_Hours_Per_Week'] >= 0]
result_data = result_data[result_data['Cholesterol_Level'] >= 0]
result_data = result_data[result_data['Blood_Pressure'] >= 0]

# 3. using z score to remove outliers
numeric_columns = data.select_dtypes(include=['number']).columns

# Calculate the z-score for each student's height
for column in numeric_columns:
    z = np.abs(stats.zscore(result_data[column]))
    
    # Identify outliers ass students with a z-score greater than 3
    threshold = 3
    outliers = result_data[z > threshold]

    result_data = result_data.drop(outliers.index)

# Reset data frame index
result_data.reset_index(drop=True, inplace=True)

result_data

Age  Income  Gender          Education  Stress_Level  \
0      50 3093457    Male  Bachelor's Degree             5   
1      44 5545445    Male  Bachelor's Degree             6   
2      25 4401808    Male    Master's Degree             6   
3      41 4606865  Female        High School             6   
4      24 6956049    Male  Bachelor's Degree             6   
...   ...     ...     ...                ...           ...   
1912   22 2072546    Male    Master's Degree             5   
1913   25 7456909  Female    Master's Degree             5   
1914   44 5560391    Male                PhD             7   
1915   20 3436759    Male  Bachelor's Degree             6   
1916   51 3343102    Male  Bachelor's Degree             6   

      Exercise_Hours_Per_Week  Cholesterol_Level  Weight  Height  \
0                          17                200      71     150   
1                          13                201      71     149   
2                          11                200      67     150   
3                           9                200      68     150   
4                          10                199      80     150   
...                       ...                ...     ...     ...   
1912                       14                201      50     150   
1913                        5                199      74     150   
1914                        4                201      44     151   
1915                        6                200      85     151   
1916                       12                200      86     151   

      Blood_Pressure Health_Status  
0                126     Excellent  
1                126          Fair  
2                127     Excellent  
3                117          Poor  
4                114          Fair  
...              ...           ...  
1912             114          Fair  
1913             108     Excellent  
1914             138          Poor  
1915             128          Good  
1916             131          Poor  

[1917 rows x 11 columns]

## Nomor 3

Membuat Visualisasi plot distribusi. Berikan uraian penjelasan kondisi setiap kolom
berdasarkan kedua plot tersebut. Jika numerik dapat dibuat dalam bentuk histogram dan
box plot, dan jika string dengan histogram.

## Nomor 4

Menentukan distribusi setiap kolom numerik menggunakan hasil visualisasi histogram.
Apakah kolom tersebut berdistribusi normal? Jika bukan, terdistribusi seperti apa kolom
tersebut?

# Soal khusus dataset

Mira adalah seorang peneliti kesehatan yang bekerja di sebuah lembaga riset medis
yang terkemuka. Sebagai bagian dari tugasnya, Mira memiliki akses ke sebuah dataset yang
berisi informasi tentang profil kesehatan dan gaya hidup dari sejumlah individu. Mira
bertanggung jawab untuk melakukan analisis statistika terhadap dataset ini guna mendapatkan
pemahaman yang lebih baik tentang faktor-faktor yang mempengaruhi kesehatan dan kualitas
hidup individu. Selain itu, Mira juga diminta untuk mengidentifikasi pola dan hubungan yang
signifikan antara variabel-variabel tersebut, serta untuk menjawab berbagai pertanyaan
penelitian yang diajukan oleh lembaga riset.


Atribut: Age, Income, Gender, Education, Stress_Level, Exercise_Hours_Per_Week,
Cholesterol_Level, Weight,Height, Blood_Pressure, Health_Status


Gunakan alpha = 0.05

## Nomor 5: Hipotesis 1 sampel

- Lembaga riset saat ini sedang mempertanyakan data berat badan individu yang
disimpan untuk kepentingan riset lanjutan. Identifikasilah apakah rata-rata berat badan
pasien diatas 65 kg?
- Tekanan darah sistole yang normal berada pada rentang 120 mmHg. Lembaga riset
perlu untuk memastikan apakah data individu yang diukur cukup normal. Periksalah
apakah rata-rata tekanan darah sistole bernilai 120 mmHg?
- Periksalah apakah data 200 individu pertama pengujian (baris teratas) memiliki rata-rata
waktu olahraga per minggu tidak sama dengan 15 jam?
- Apakah penduduk dengan pendapatan yang lebih besar dari Rp 7.500.000,00 tidak
sama dengan 30% dari data keseluruhan individu?

## Nomor 6: Hipotesis 2 sampel

Lembaga riset membagi data individu menjadi dua bagian, yaitu data individu yang lebih
awal masuk data penelitian (bagian atas) dan yang baru saja (bagian bawah).

- Periksa apakah rata-rata berat badan individu yang lebih awal masuk data penelitian
sama dengan rata-rata berat badan individu yang masuk baru saja?
- Bagaimana dengan pendapatan individu, apakah pendapatan sistole individu yang lebih
awal masuk data penelitian lebih besar Rp 1.250.000,00 dari yang baru saja masuk?
- Lembaga riset ingin membandingkan kondisi kesehatan individu dari dua bagian data.
Apakah variansi tekanan darah individu yang lebih awal masuk data penelitian sama
dengan yang baru saja masuk?
- Apakah proporsi waktu olahraga yang lebih dari 8 jam per minggu pada data individu
awal lebih besar daripada kuantitas proporsi pada data individu akhir dengan waktu
olahraga yang sama?